pkg/scm/git/git.go

```
@@ -537,7 +570,66 @@ func (h *stiGit) SubmoduleUpdate(repo string, init, recursive bool) error {
 		Stderr: os.Stderr,
 		Dir:    repo,
 	}
-	return h.runner.RunWithOptions(opts, "git", updateArgs...)
+	return h.RunWithOptions(opts, "git", updateArgs...)
+}
+
+// LsTree returns a slice of os.FileInfo objects populated with the paths and
+// file modes of files known to Git.  This is used on Windows systems where the
+// executable mode metadata is lost on git checkout.
+func (h *stiGit) LsTree(repo, ref string, recursive bool) ([]os.FileInfo, error) {
+	rx := regexp.MustCompile("([0-7]{6}) [^ ]+ [0-9a-f]{40}\t(.*)")
+
+	args := []string{"ls-tree", ref}
+	if recursive {
+		args = append(args, "-r")
+	}
+
+	opts := util.CommandOpts{
+		Dir: repo,
+	}
+
+	r, err := h.StartWithStdoutPipe(opts, "git", args...)
+	if err != nil {
+		return nil, err
+	}
+
+	submodules := []string{}
+	rv := []os.FileInfo{}
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		text := scanner.Text()
+		m := rx.FindStringSubmatch(text)
+		if m == nil {
+			return nil, fmt.Errorf("unparsable response %q from git ls-files", text)
+		}
+		mode, _ := strconv.ParseInt(m[1], 8, 0)
```

@jim-minter, can we safely ignore the error here?